File: copy_project.mjs
import fs from 'fs';
import path from 'path';

const SOURCE_DIR = 'd:\\1V\\ailegalarmenia-main';
const DEST_DIR = 'd:\\1V\\NEW_REPO_FOLDER_NAME';

const EXCLUDE_DIRS = new Set([
    '.git', 'node_modules', 'dist', 'build', '.next', '.turbo', '.cache', 
    'coverage', '__pycache__', '.venv', 'venv'
]);

const EXCLUDE_FILES_REGEX = [
    /^\.env(\..*)?$/,
    /^\.DS_Store$/,
    /^copy_project\.(py|mjs)$/,
    /.*\.log$/,
    /^Thumbs\.db$/
];

const SKIP_SECRET_SCAN_REGEX = [
    /.*\.lock$/,
    /^package-lock\.json$/,
    /.*\.svg$/,
    /.*\.txt$/,
    /^skills-lock\.json$/,
    /.*\.png$/,
    /.*\.jpg$/,
    /.*\.woff2?$/
];

const SECRET_REGEX = [
    /api[_-]?key\s*[:=]\s*["']?[a-zA-Z0-9_\-]{20,}["']?/i,
    /secret\s*[:=]\s*["']?[a-zA-Z0-9_\-]{20,}["']?/i,
    /token\s*[:=]\s*["']?[a-zA-Z0-9_\-]{20,}["']?/i,
    /password\s*[:=]\s*["']?[a-zA-Z0-9_\-]{8,}["']?/i,
    /-----BEGIN (RSA )?PRIVATE KEY-----/
];

const security_findings = [];
const copied_files = [];
const excluded_items = [];

if (fs.existsSync(DEST_DIR)) {
    fs.rmSync(DEST_DIR, { recursive: true, force: true });
}
fs.mkdirSync(DEST_DIR, { recursive: true });

function walkSync(currentDirPath) {
    const items = fs.readdirSync(currentDirPath, { withFileTypes: true });
    for (const item of items) {
        const fullPath = path.join(currentDirPath, item.name);
        const relPath = path.relative(SOURCE_DIR, fullPath);
        
        if (item.isDirectory()) {
            if (EXCLUDE_DIRS.has(item.name)) {
                excluded_items.push(relPath + '\\');
                continue;
            }
            fs.mkdirSync(path.join(DEST_DIR, relPath), { recursive: true });
            walkSync(fullPath);
        } else {
            if (EXCLUDE_FILES_REGEX.some(regex => regex.test(item.name))) {
                excluded_items.push(relPath);
                continue;
            }
            
            const destPath = path.join(DEST_DIR, relPath);
            let skip_copy = false;
            
            if (!SKIP_SECRET_SCAN_REGEX.some(regex => regex.test(item.name))) {
                try {
                    const content = fs.readFileSync(fullPath, 'utf-8');
                    for (const pattern of SECRET_REGEX) {
                        if (pattern.test(content)) {
                            security_findings.push(`${relPath} contains potential secret`);
                            skip_copy = true;
                            break;
                        }
                    }
                } catch (e) {
                    // Ignore read errors, mostly for binary files
                }
            }
            
            if (skip_copy) {
                excluded_items.push(`${relPath} (Secret found)`);
                continue;
            }
            
            fs.copyFileSync(fullPath, destPath);
            copied_files.push(relPath);
        }
    }
}

walkSync(SOURCE_DIR);

const gitignoreContent = `# Dependencies
node_modules/
.venv/
venv/
__pycache__/

# Builds
dist/
build/
.next/
.turbo/
out/

# Caches
.cache/
coverage/
.eslintcache

# Environment
.env
.env.*

# Logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# System
.DS_Store
Thumbs.db
`;

fs.writeFileSync(path.join(DEST_DIR, '.gitignore'), gitignoreContent, 'utf-8');

fs.writeFileSync('d:\\1V\\copy_log.txt', 
    "Security findings:\n" + security_findings.join("\n") + "\n\n" +
    "Excluded items:\n" + excluded_items.join("\n") + "\n\n" +
    "Copied files count: " + copied_files.length
);

File: copy_project.py
import os
import shutil
import re

SOURCE_DIR = r"d:\1V\ailegalarmenia-main"
DEST_DIR = r"d:\1V\NEW_REPO_FOLDER_NAME"

EXCLUDE_DIRS = {
    '.git', 'node_modules', 'dist', 'build', '.next', '.turbo', '.cache', 
    'coverage', '__pycache__', '.venv', 'venv'
}

EXCLUDE_FILES_REGEX = [
    re.compile(r'^\.env(\..*)?$'),
    re.compile(r'^\.DS_Store$'),
    re.compile(r'^copy_project\.py$'),
    re.compile(r'.*\.log$'),
    re.compile(r'^Thumbs\.db$')
]

SKIP_SECRET_SCAN_REGEX = [
    re.compile(r'.*\.lock$'),
    re.compile(r'^package-lock\.json$'),
    re.compile(r'.*\.svg$'),
    re.compile(r'.*\.txt$'),
    re.compile(r'^skills-lock\.json$')
]

SECRET_REGEX = [
    r'(?i)api[_-]?key\s*[:=]\s*["\']?[a-zA-Z0-9_\-]{20,}["\']?',
    r'(?i)secret\s*[:=]\s*["\']?[a-zA-Z0-9_\-]{20,}["\']?',
    r'(?i)token\s*[:=]\s*["\']?[a-zA-Z0-9_\-]{20,}["\']?',
    r'(?i)password\s*[:=]\s*["\']?[a-zA-Z0-9_\-]{8,}["\']?',
    r'-----BEGIN (?:RSA )?PRIVATE KEY-----'
]

security_findings = []
copied_files = []
excluded_items = []

if os.path.exists(DEST_DIR):
    shutil.rmtree(DEST_DIR)
os.makedirs(DEST_DIR)

for root, dirs, files in os.walk(SOURCE_DIR):
    original_dirs = list(dirs)
    dirs[:] = [d for d in dirs if d not in EXCLUDE_DIRS]
    
    for d in original_dirs:
        if d in EXCLUDE_DIRS:
            rel = os.path.relpath(os.path.join(root, d), SOURCE_DIR)
            excluded_items.append(rel + "/")
            
    rel_path = os.path.relpath(root, SOURCE_DIR)
    dest_path = os.path.join(DEST_DIR, rel_path) if rel_path != '.' else DEST_DIR
    
    if not os.path.exists(dest_path):
        os.makedirs(dest_path)
        
    for f in files:
        if any(regex.match(f) for regex in EXCLUDE_FILES_REGEX):
            excluded_items.append(os.path.join(rel_path, f))
            continue
            
        src_file = os.path.join(root, f)
        dest_file = os.path.join(dest_path, f)
        
        if not any(regex.match(f) for regex in SKIP_SECRET_SCAN_REGEX):
            try:
                is_secret = False
                with open(src_file, 'r', encoding='utf-8') as f_in:
                    content = f_in.read()
                    for pattern in SECRET_REGEX:
                        matches = re.findall(pattern, content)
                        if matches:
                            is_secret = True
                            security_findings.append(f"{os.path.join(rel_path, f)} contains potential secret")
                            break
                if is_secret:
                    excluded_items.append(os.path.join(rel_path, f) + " (Secret found)")
                    continue
            except UnicodeDecodeError:
                pass
            
        shutil.copy2(src_file, dest_file)
        copied_files.append(os.path.join(rel_path, f))

# Write updated .gitignore
gitignore_content = """# Dependencies
node_modules/
.venv/
venv/
__pycache__/

# Builds
dist/
build/
.next/
.turbo/
out/

# Caches
.cache/
coverage/
.eslintcache

# Environment
.env
.env.*

# Logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# System
.DS_Store
Thumbs.db
"""
with open(os.path.join(DEST_DIR, '.gitignore'), 'w', encoding='utf-8') as f:
    f.write(gitignore_content)

with open(r'd:\1V\copy_log.txt', 'w', encoding='utf-8') as log_out:
    log_out.write("Security findings:\n" + "\n".join(security_findings) + "\n\n")
    log_out.write("Excluded items:\n" + "\n".join(excluded_items) + "\n\n")
    log_out.write("Copied files count: " + str(len(copied_files)))
